Feat: Add Support for TWRNC

abhishek310001 opened the pull request:
## Description
_This PR introduces the integration of `twrnc` into the existing boilerplate to use Tailwind CSS._

## Screenshots:

| ![Screenshot_1718008097](https://github.com/jalantechnologies/boilerplate-react-native/assets/128154648/3c3032f3-6f89-4420-a53a-04ceab96a4f5) | ![Screenshot_1718008104](https://github.com/jalantechnologies/boilerplate-react-native/assets/128154648/28ee63fd-b9d9-41a7-b8b2-7d46bc69cb8e) | ![Screenshot_1718008117](https://github.com/jalantechnologies/boilerplate-react-native/assets/128154648/b68c6ece-2078-4442-9a01-4d08d6f9e7d4) |
| ---- | ---- | ---- |
| ![Screenshot_1718008100](https://github.com/jalantechnologies/boilerplate-react-native/assets/128154648/ccf33857-738f-477f-af6c-9d334e7d2385) | ![Screenshot_1718008107](https://github.com/jalantechnologies/boilerplate-react-native/assets/128154648/3f52b4c9-cf3f-422d-8441-618d956f66b5) | ![Screenshot_1718008120](https://github.com/jalantechnologies/boilerplate-react-native/assets/128154648/7737dd96-3536-4635-9d75-957bbaa8110a) |
| ![Simulator Screenshot - iPhone SE (3rd generation) - 2024-06-10 at 13 57 09](https://github.com/jalantechnologies/boilerplate-react-native/assets/128154648/8f9e246c-0e10-4460-9a46-93c4abfda566) | ![Simulator Screenshot - iPhone SE (3rd generation) - 2024-06-10 at 13 57 17](https://github.com/jalantechnologies/boilerplate-react-native/assets/128154648/2605883c-ddb1-44bd-b451-c2ab9922c9c8) | ![Simulator Screenshot - iPhone SE (3rd generation) - 2024-06-10 at 13 57 58](https://github.com/jalantechnologies/boilerplate-react-native/assets/128154648/c4e8ad7e-3a33-498a-8893-38aa04672ccd) |
| ![Simulator Screenshot - iPhone SE (3rd generation) - 2024-06-10 at 13 57 11](https://github.com/jalantechnologies/boilerplate-react-native/assets/128154648/ecbe29dd-4f8d-4603-a61f-987dd61619b7) | ![Simulator Screenshot - iPhone SE (3rd generation) - 2024-06-10 at 13 57 38](https://github.com/jalantechnologies/boilerplate-react-native/assets/128154648/2309a98b-a4fa-4659-b7b2-68f86c8023f7)| ![Simulator Screenshot - iPhone SE (3rd generation) - 2024-06-10 at 13 58 01](https://github.com/jalantechnologies/boilerplate-react-native/assets/128154648/f97597ef-2d7e-4257-a87f-3fb5e180cf04) |


## Videos:

[twrnc.webm](https://github.com/jalantechnologies/boilerplate-react-native/assets/128154648/3587eb7b-08e0-438b-a900-f493e9d6404e)

https://github.com/jalantechnologies/boilerplate-react-native/assets/128154648/22805e92-c30a-490b-9f34-2c4ec2065f3c





## Database schema changes
NA

## Tests
### Automated test cases added
NA

### Manual test cases run
_For each manual test case, list the steps to test or reproduce the PR._
- Tested on following devices:
- [x] iPhoneSE 3rd Gen (iOS 17.2)
- [x] Pixel 3a (Android 14)

abhishek310001 committed:
added nativewind support

abhishek310001 committed:
Update test for updated style

Patel-world commented on src/components/custom-button/custom-button.tsx:
```diff
@@ -1,6 +1,6 @@
 import { Button } from '@rneui/themed';
 import React from 'react';
-import { useStyles } from './styles';
+import { useTailwind } from 'boilerplate-react-native/src/utils/tailwind-util';
```
relative path

abhishek310001 replied:
Updated!

Patel-world commented on src/screens/home/home.tsx:
```diff
@@ -1,22 +1,21 @@
 import { CustomButton } from 'boilerplate-react-native/src/components';
-import { useStyles } from './styles';
 import { useTranslation } from 'react-i18next';
+import { useTailwind } from 'boilerplate-react-native/src/utils/tailwind-util';
```
relative path here too and take care of other places

abhishek310001 replied:
Sure, I will try to use relative path.

abhishek310001 committed:
chore: Update import paths for CustomButton and useTailwind in custom-button.tsx and home.tsx

Patel-world reviewed (approved):
LGTM

abhishek310001 committed:
deleted pascal case files

abhishek310001 committed:
Merge branch 'main' into abhishek/feat/issue-23-integrate-nativewind

abhishek310001 committed:
Revert "deleted pascal case files"

This reverts commit bb87731ba5999f3bc6700baa4dd07803dc4af45e.

abhishek310001 committed:
updated ios project

abhishek310001 committed:
update Application.tsx

codVenturer commented on tailwind.config.js:
```diff
@@ -0,0 +1,233 @@
+        success: '#219653',
+        danger: '#D34053',
+        warning: '#FFA70B',
+      },
```
Isn't this all available by default, why do you need to extend it?
just add a few examples of extend the theme. that would be good enough

abhishek310001 replied:
The same file was used for MERN boilerplate so I just replicated that

jjalan replied:
These are pretty poor reasoning. We need to be very thoughtful, and careful about each line of code that we checkin in boilerplate as it has affect across all future applications. Please avoid this in future.

If we feel there is a technical debt on other boilerplate, please open a JIRA ticket on it.

abhishek310001 replied:
Noted

abhishek310001 committed:
lint fix

codVenturer commented on src/utils/tailwind-util.ts:
```diff
@@ -0,0 +1,6 @@
+import { StyleSheet, StyleProp, ViewStyle } from 'react-native';
+import { useTailwind as tailwind } from 'nativewind';
+
+export const useTailwind = (className: string): StyleProp<ViewStyle> => {
```
every style won't be ViewStyle, there are more of them.
Also seems like `useTailwind` has been [removed](https://www.nativewind.dev/guides/tailwindcss-react-native#usetailwind)

abhishek310001 replied:
You are right not every style will be ViewStyle, so this has to updated as per the usage since this is the only alternative I've got for using tailwind classes on other styles prop of a component.
> Also seems like useTailwind has been [removed](https://www.nativewind.dev/guides/tailwindcss-react-native#usetailwind)

Somehow it is working as of now

jjalan replied:
> Somehow it is working as of now

This is a *very* loose statement and often becomes hard to debug or cause unexpected behaviour in application. In general, these are pretty poor reasoning. We need to be very thoughtful, and careful about each line of code that we checkin in boilerplate as it has affect across all future applications. Please avoid this in future.

abhishek310001 committed:
chore: Remove unused tailwind classes from custom-button.tsx

abhishek310001 committed:
feat: Update primary and secondary colors in tailwind.config.js

The primary and secondary colors in the `tailwind.config.js` file have been updated to use light and dark shades. This change provides more flexibility in styling the application.

abhishek310001 committed:
added custom input and button using tailwind

abhishek310001 committed:
chore: Update home screen with custom input and button components

abhishek310001 committed:
modified paths for components

abhishek310001 committed:
chore: Remove unused tailwind classes from custom-button.tsx

codVenturer commented on src/components/custom-button/tailwind-button.tsx:
```diff
@@ -0,0 +1,56 @@
+}) => {
+  const buttonStyle = [
+    'px-2.5',
+    'py-1.5',
```
This fails the purpose of adding a Component library, I don't think this would work. 
There are places in the components from react itself that have the props `contentContainerStyle`, `headerStyle` and all like in Flatlist. how are we going to implement in those.
and if there is need for some new components like google maps, I don't think you can do this.
Please try to think about those scenarios and come up with something robust.
I'll suggest you to look at some open source apps created with NativeWind and see what they do to mitigate the above issues.

abhishek310001 replied:
Hey,
Looked at multiple open source apps with `nativewind` integration and found out that these styles are not handled with tailwind instead it is used normally with either stylesheet or inline styles.
The best alternative I could find to apply tailwind CSS is using [twrnc](https://www.npmjs.com/package/twrnc?activeTab=readme).
It can be used with most of the style props using tailwind CSS.

abhishek310001 committed:
chore: Remove nativewind and add twrnc

abhishek310001 committed:
update Application.tsx
